process/track/baseBlockTrack.go

```diff
@@ -0,0 +1,425 @@
+func (bbt *baseBlockTrack) GetLastCrossNotarizedHeader(shardID uint32) (data.HeaderHandler, []byte, error) {
+	return bbt.crossNotarizer.getLastNotarizedHeader(shardID)
+}
+
+// GetLastHeader returns the last header received (highest round) for the given shard
+func (bbt *baseBlockTrack) GetLastHeader(shardID uint32) data.HeaderHandler {
+	bbt.mutHeaders.RLock()
+	defer bbt.mutHeaders.RUnlock()
+
+	var lastHeaderForShard data.HeaderHandler
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return lastHeaderForShard
+	}
+
+	maxRound := uint64(0)
+	for _, headersInfo := range headersForShard {
+		for _, headerInfo := range headersInfo {
+			if headerInfo.header.GetRound() > maxRound {
+				maxRound = headerInfo.header.GetRound()
+				lastHeaderForShard = headerInfo.header
+			}
+		}
+	}
+
+	return lastHeaderForShard
+}
+
+// GetTrackedHeadersForShard returns tracked headers for a given shard
+func (bbt *baseBlockTrack) GetTrackedHeaders(shardID uint32) ([]data.HeaderHandler, [][]byte) {
+	return bbt.sortHeadersFromNonce(shardID, 0)
+}
+
+func (bbt *baseBlockTrack) sortHeadersFromNonce(shardID uint32, nonce uint64) ([]data.HeaderHandler, [][]byte) {
+	bbt.mutHeaders.RLock()
+	defer bbt.mutHeaders.RUnlock()
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return nil, nil
+	}
+
+	sortedHeadersInfo := make([]*headerInfo, 0)
+
+	for headersNonce, headersInfo := range headersForShard {
+		if headersNonce < nonce {
+			continue
+		}
+
+		sortedHeadersInfo = append(sortedHeadersInfo, headersInfo...)
+	}
+
+	if len(sortedHeadersInfo) > 1 {
+		sort.Slice(sortedHeadersInfo, func(i, j int) bool {
+			return sortedHeadersInfo[i].header.GetNonce() < sortedHeadersInfo[j].header.GetNonce()
+		})
+	}
+
+	headers := make([]data.HeaderHandler, 0)
+	headersHashes := make([][]byte, 0)
+
+	for _, headerInfo := range sortedHeadersInfo {
+		headers = append(headers, headerInfo.header)
+		headersHashes = append(headersHashes, headerInfo.hash)
+	}
+
+	return headers, headersHashes
+}
+
+// GetTrackedHeadersWithNonce returns tracked headers for a given shard and nonce
+func (bbt *baseBlockTrack) GetTrackedHeadersWithNonce(shardID uint32, nonce uint64) ([]data.HeaderHandler, [][]byte) {
+	bbt.mutHeaders.RLock()
+	defer bbt.mutHeaders.RUnlock()
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return nil, nil
+	}
+
+	headersForShardWithNonce, ok := headersForShard[nonce]
+	if !ok {
+		return nil, nil
+	}
+
+	headers := make([]data.HeaderHandler, 0)
+	headersHashes := make([][]byte, 0)
+
+	for _, headerInfo := range headersForShardWithNonce {
+		headers = append(headers, headerInfo.header)
+		headersHashes = append(headersHashes, headerInfo.hash)
+	}
+
+	return headers, headersHashes
+}
+
+// IsShardStuck returns true if the given shard is stuck
+func (bbt *baseBlockTrack) IsShardStuck(shardId uint32) bool {
+	header := bbt.GetLastHeader(shardId)
+	if check.IfNil(header) {
+		return false
```

Comment:
if there is no last header means the shard is ok ? I think it should return true

Reply:
This method is used in fork detector and could be used anywhere to determine if one cross chain is stuck or not. Actually GetLastHeader could not be nil as it is initialized with startHeaders(genesis header for each shard) on New.. method. But even if it would be nil for some reason, you can not say that a shard is stuck just because you have no info about it. This is why it returns false in this case, but again this would not be possible with the current implementation.